### PR TITLE
Switch UnitSelector over to use course versions

### PR DESCRIPTION
**Pull request by dmcavoy**
We use the UnitSelector component in the TeacherDashboard on the Progress, TextResponse and Assessments tabs to allow a teacher to pick which unit they want to see information for. This dropdown was being set up with information that still came from ScriptConstants which we want to move off of to make launching curriculum self serve.  

This PR moves the UnitSelector over to using CourseVersion and Unit to get the information needed to populate the dropdown. It leverages the summary work we just did for the AssignmentSelector but only gives CourseVersions and their Units which have progress.

One difference to be noted here is that all the CSF courses for a particular year are not grouped together in the list. @tess323 @fontie715 are we okay with this change?

Screen Shot 2022-04-09 at 12 22 07 AM

## Links

[ENG plan](https://docs.google.com/document/d/1V61xONU0-mleMEEdHWNxhvZtYIPRCB31mT_hbUXusCQ/edit#)
[Product Spec](https://docs.google.com/document/d/1qZpc90daxQiiqpcQUiUiGV2tQoaYEQUzOA72UZzBlVE/edit#heading=h.7j0u06e8a26f)
[Jira](https://codedotorg.atlassian.net/browse/PLAT-1685)

## Testing story

- Updated unit tests
- Updated UI tests
- Manual testing comparing before and after

### CODE REVIEW

**Comment by dmcavoy** [comment]
UI tests - give levelbuilder permission to see hidden things?

**Comment by dmcavoy** [comment]
Make sure course version is assignable. don't want a pilot script that a student is in for another teacher to show

File: apps/src/redux/unitSelectionRedux.js
Comment (dmcavoy) [inline apps/src/redux/unitSelectionRedux.js:94]: Check what happens if there are no scripts with progress

**Comment by dmcavoy** [comment]
Test to check that sending unit id if no scriptId provided unitSelectionRedux

**Comment by dmcavoy** [comment]
For UI Tests: The section that gets assigned the course is not the same section as one with the student and teacher associated to each other

**Comment by tess323** [comment]
I will defer to @fontie715 here.  For me, it's not ideal, but it's at least not confusing.  Would all of those course versions only appear if students currently in the section had made progress in them?

**Comment by dmcavoy** [comment]
>I will defer to @fontie715 here. For me, it's not ideal, but it's at least not confusing. Would all of those course versions only appear if students currently in the section had made progress in them?

Yes they only show up if someone in the section has progress in them

File: apps/src/redux/unitSelectionRedux.js
Comment (jamescodeorg) [inline apps/src/redux/unitSelectionRedux.js]: (minor) I think it might be clearer to put all of the logic for calculating `scriptId` together here (pulling up the conditional that's currently on lines 63-65).

File: apps/src/templates/sectionProgress/sectionProgressRedux.js
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/sectionProgressRedux.js]: Did you intend to leave these log statements in?

File: apps/src/templates/sectionProgress/sectionProgressTestHelpers.js
Comment (jamescodeorg) [inline apps/src/templates/sectionProgress/sectionProgressTestHelpers.js]: Just curious, did you change this because 162 didn't work?

File: apps/src/redux/unitSelectionRedux.js
Comment (jamescodeorg) [inline apps/src/redux/unitSelectionRedux.js]: I'll note that I initially thought "course versions with progress" meant an array of objects with course version and progress information mashed together instead of an array of course version objects where students have made progress.  I can't think of anything that's clearer though.

File: apps/test/unit/redux/unitSelectionReduxTest.js
Comment (jamescodeorg) [inline apps/test/unit/redux/unitSelectionReduxTest.js]: I'm a little worried that the relationship between the fake course version data and the unit test is a little unclear.  For example, why is the expected value here '1'?  If I change the fake course version data in teacherDashboardTestHelpers, would I expect the tests here to need to be updated?

Ideally, we would craft the minimal state needed in each test.  If that's too much work for now, we can tackle it in the future.

File: dashboard/app/models/course_version.rb
Comment (jamescodeorg) [inline dashboard/app/models/course_version.rb]: This API was a little confusing to me.  I think it's because, from the name, I expected it to look at progress and the parameter name `student_unit_ids` doesn't indicate that it has to do with progress.

I'm wondering if maybe this API can be something more generic like `course_versions_containing_units` and just take an array of `unit_ids`?

File: dashboard/app/models/course_version.rb
Comment (jamescodeorg) [inline dashboard/app/models/course_version.rb]: (Same comment as above.)

**Review by jamescodeorg — COMMENTED** [review]
Phew!  Thanks for tackling this very gnarly problem!

**Review by jamescodeorg — COMMENTED** [review]
Side question: I continue to find the term "CourseVersion" kind of confusing.  Reading through this review, I think a good name for this concept would be "Course" but that was previously used to mean something else.  Do you think that by the time we go to clean the top of the curriculum model hierarchy up, will the term "Course" be available enough for us to use it for this concept?

File: apps/src/templates/sectionProgress/sectionProgressTestHelpers.js
Comment (dmcavoy) [reply apps/src/templates/sectionProgress/sectionProgressTestHelpers.js]: Yes - The ID has to match what is in the unit selector dropdown and the fake things we seed the unit selector dropdown with defaults to 1 now

Comment (jamescodeorg) [reply apps/src/templates/sectionProgress/sectionProgressTestHelpers.js]: Oh, I see.  I don't know how much work it would be, but it would be nice if the tests were more self-contained and the data was defined here so there's not coupling between different unit tests.

**Comment by dmcavoy** [comment]
Here is an alternative to the one dropdown. Trying to recreate the dropdown exactly as it exists right now does not really match our data model so we will likely need to consider some kind of change here. 

Screen Shot 2022-04-14 at 11 30 08 PM

PR: https://github.com/code-dot-org/code-dot-org/pull/45849 with code for this which is not ready for review

File: apps/src/templates/sectionProgress/sectionProgressTestHelpers.js
Comment (dmcavoy) [reply apps/src/templates/sectionProgress/sectionProgressTestHelpers.js]: This was the data I had to update for the Storybook for ProgressTableView so I'm not planning to work on cleaning it up as part of this work.

**Comment by dmcavoy** [comment]
Talking to the PM team we are planning to go forward with this approach with one update. We will group all version years of the same course offering together if the course versions have a content root of type script. Example: All Course A's will be under the Course A header. I will make those updates and response to PR feedback and request re-review when completed.

File: apps/src/redux/unitSelectionRedux.js
Comment (dmcavoy) [reply apps/src/redux/unitSelectionRedux.js]: Do you like courseVersionsForSection better?

File: apps/test/unit/redux/unitSelectionReduxTest.js
Comment (dmcavoy) [reply apps/test/unit/redux/unitSelectionReduxTest.js]: I agree with you that using the fake data here is unclear. I was working to match the existing pattern that was already being used in this file. I think with the timeline of this work I'm going to hold off on trying to update this right now but agree it would be a good thing to come back to.

File: apps/src/redux/unitSelectionRedux.js
Comment (jamescodeorg) [reply apps/src/redux/unitSelectionRedux.js]: Not really, thanks for thinking about this though!

File: apps/test/unit/redux/unitSelectionReduxTest.js
Comment (jamescodeorg) [reply apps/test/unit/redux/unitSelectionReduxTest.js]: Could you open a ticket on either EPTools or LP (wherever you think this more appropriately belongs) to return to this?

**Comment by dmcavoy** [comment]
Screen Shot 2022-04-18 at 9 36 40 PM

File: apps/test/unit/redux/unitSelectionReduxTest.js
Comment (dmcavoy) [reply apps/test/unit/redux/unitSelectionReduxTest.js]: https://codedotorg.atlassian.net/browse/LP-2325

**Comment by dmcavoy** [comment]
> Side question: I continue to find the term "CourseVersion" kind of confusing. Reading through this review, I think a good name for this concept would be "Course" but that was previously used to mean something else. Do you think that by the time we go to clean the top of the curriculum model hierarchy up, will the term "Course" be available enough for us to use it for this concept?

The confusion here is something we definitely want to work on improving. There are a bunch of things we are considering to improve this area and it will depend on time allocated for the work what we do I think. I will definitely keep this in mind as we plan

File: apps/test/unit/redux/unitSelectionReduxTest.js
Comment (jamescodeorg) [reply apps/test/unit/redux/unitSelectionReduxTest.js]: Thanks!

File: lib/cdo/script_constants.rb
Comment (davidsbailey) [inline lib/cdo/script_constants.rb:398]: so glad to see this go!

**Review by davidsbailey — APPROVED** [review]
nice work Dani! I mostly looked at the backend code. LGTM once James's concerns are addressed.

**Comment by dmcavoy** [comment]
@jamescodeorg if you want to take one more look can you do that today so I can merge this in tomorrow?

File: apps/src/redux/unitSelectionRedux.js
Comment (jamescodeorg) [inline apps/src/redux/unitSelectionRedux.js:17]: (unrelated to this change)  I know we're midway through a rename but this file in particular seems to have a mix of script and unit terminology.  Mostly curious, what's our plan to "finish" the rename?  Any sense of what the ends state looks like (e.g. what parts of the system will still use the script terminology?)?

File: dashboard/app/models/unit_group.rb
Comment (jamescodeorg) [inline dashboard/app/models/unit_group.rb:365]: (minor) Does a unit group contain units or vice versa?  (If unit groups contain units, maybe `contains_units?` would be a clearer name?)

File: dashboard/app/models/course_version.rb
Comment (jamescodeorg) [inline dashboard/app/models/course_version.rb]: I think some comments here describing the algorithm for assembling the list would be helpful.  Include some background on _why_ single unit course offerings are handled differently than unit group course versions.

Btw, why do we ask CourseOffering for single unit courses but CourseVersion for unit group courses?

Comment (jamescodeorg) [inline dashboard/app/models/course_version.rb]: Why does this method depend on user?  Do we want to hide courses in the section progress dashboard that the teacher does not have access to?  Would we consider showing them if it simplifies our logic here?

For the future, I think it would be helpful to pull the filtering by user "up" as far as possible so that as few methods as possible need the user as input.  (For this particular instance, another option would be to filter by unit_ids first and only pass the list of "acceptable" unit ids to these various functions.)

File: dashboard/app/models/course_version.rb
Comment (jamescodeorg) [inline dashboard/app/models/course_version.rb]: (minor) For the future, I think it's helpful to be able to look at a file and quickly locate the main intended entry points.  One way this can be done is by marking helper methods such as this one as private or moving all of the helper methods to a private section.

**Review by jamescodeorg — APPROVED** [review]
Thanks for the last round of changes.  I had a few more minor questions and suggestions on how to structure some of the code in the future, but nothing that blocks this PR.  This was a ton of work, way to push through!

File: apps/src/redux/unitSelectionRedux.js
Comment (davidsbailey) [reply apps/src/redux/unitSelectionRedux.js:17]: > (unrelated to this change) I know we're midway through a rename but this file in particular seems to have a mix of script and unit terminology. Mostly curious, what's our plan to "finish" the rename? 

apologies... this has been forever in my personal backlog. we accidentally started this work at a time when we didn't explicitly make time to finish it. the goal is to finish it this Summer. 

> Any sense of what the ends state looks like (e.g. what parts of the system will still use the script terminology?)?

* in the database, the `scripts` table and various db fields named `script_id` are likely to remain unchanged
* routing will remain on /s/ for now. it's unclear whether controllers will need to remain named scripts_controller in order to keep these routes working
* everything else in dashboard and apps code should change to `unit`

Comment (dmcavoy) [reply apps/src/redux/unitSelectionRedux.js:17]: I think we are trying to move as much as we can over to use unit. I generally try to move things as I touch them. I stopped short of changing some things like scriptId in this PR because I didn't want the size of it to grow. I think the only things we would not update would be things that are breaking to change such as translation strings. I don't know where we are with the work getting scheduled. @davidsbailey might know more

File: dashboard/app/models/unit_group.rb
Comment (dmcavoy) [reply dashboard/app/models/unit_group.rb:365]: Ya I didn't love this name for unit_group but I wanted to use the same name for the unit_group method as the script method so that course_version did not need special casing. I was having a hard time thinking of a better name. The method used to be 'has_student_progress' but when I tried to make things more generic I landed here. I'm open to suggestions of something that would work for both unit and unit_group

Comment (dmcavoy) [reply dashboard/app/models/unit_group.rb:365]: I guess it could be contains_units? a unit kind of contains itself

File: dashboard/app/models/course_version.rb
Comment (dmcavoy) [reply dashboard/app/models/course_version.rb]: I like the idea of filter by unit_ids first and only pass the list of "acceptable" unit ids to these various functions. I can take a look at that

File: dashboard/app/models/course_version.rb
Comment (dmcavoy) [reply dashboard/app/models/course_version.rb]: Great feedback. I was wondering how to make things in this file clearer. I will updated

File: dashboard/app/models/course_version.rb
Comment (dmcavoy) [reply dashboard/app/models/course_version.rb]: Good call. I will add comments.

We use Course Offerings for single unit course offerings because we want to group together all the course offerings across years. So all the Course A's are together under the Course A header. Where as for unit group courses we want to have all the units for a specific year grouped together under the unit group for that year. So CSD has multiple headers in the list with the units for that year under it.

File: apps/src/redux/unitSelectionRedux.js
Comment (dmcavoy) [reply apps/src/redux/unitSelectionRedux.js:17]: Thanks Dave. I think I was the one that started that work when I was past my due date and looking for things to do while waiting.

File: dashboard/app/models/course_version.rb
Comment (dmcavoy) [reply dashboard/app/models/course_version.rb]: I tried to do this and it was complaining so I think I did it wrong. I'm not gonna hold the PR on this right now

**Pull request merged**
